### ISSUE TITLE
Change the way we create undirected graphs from (directed) graphs

### ISSUE THREAD

**Issue by sbromberger** [JuliaGraphs/LightGraphs.jl#853]
Right now, in connectivity.jl, we have the following code:

```
@traitfn weakly_connected_components(g::::IsDirected) = connected_components(Graph(g))
```

This works fine when you're dealing with SimpleGraphs, but it assumes you can change other graph types (like `StaticDiGraph`, for example) into `Graph`, which you can't really do.

My proposal is to create a `undirected(g::AbstractGraph)` function in core.jl that can be overridden by other packages to take a graph of a given type and return the undirected graph of the same type.

Comments / thoughts appreciated.

**Comment by sbromberger** [comment]
I think this should be part of the API instead of the core.

**Comment by sbromberger** [comment]
This actually turns out to be a larger issue (naturally). Lots of functions in `connectivity.jl` (and perhaps elsewhere) return `Graph` or `DiGraph` under the assumption that we're passing around `SimpleGraph` structures. This is now an incorrect assumption.

**Comment by sbromberger** [comment]
@lelogrott do you have time to look at this? If not, no worries; I'll unassign you.

Thanks!

**Comment by lelogrott** [comment]
@sbromberger I thought I had. I was trying to manage my time but right now it's really complicated. I'm sorry! I'll find a way of contribute. please, feel free to unassign me.

**Comment by sbromberger** [comment]
The issue with that approach is that it doesn't solve the immediate problem: that is, when you create a directed graph from an undirected graph, what graph type do you get back? When all we had were SimpleGraphs, it wasn't a problem. The issue is now that any function that makes a directed graph undirected needs to work with multiple graph types, each of which now needs a method to go from its directed version to its undirected version.

**Comment by jpfairbanks** [comment]
Closed by #931.
